source/common/protobuf/utility.cc

```diff
@@ -837,4 +837,41 @@ void StructUtil::update(ProtobufWkt::Struct& obj, const ProtobufWkt::Struct& wit
   }
 }
 
+void MessageUtil::loadFromFile(const std::string& path, Protobuf::Message& message,
+                               ProtobufMessage::ValidationVisitor& validation_visitor,
+                               Api::Api& api) {
+  auto file_or_error = api.fileSystem().fileReadToEnd(path);
+  THROW_IF_STATUS_NOT_OK(file_or_error, throw);
+  const std::string contents = file_or_error.value();
+  // If the filename ends with .pb, attempt to parse it as a binary proto.
+  if (absl::EndsWithIgnoreCase(path, FileExtensions::get().ProtoBinary)) {
+    // Attempt to parse the binary format.
+    if (message.ParseFromString(contents)) {
+      MessageUtil::checkForUnexpectedFields(message, validation_visitor);
+    }
+    return;
+  }
+
+#ifdef ENVOY_ENABLE_YAML
```

Comment:
so the original version (and the new, default-compile version) we try loading a bunch of forms based on file path, and eventually "just assume it's json"

in the new yaml-free path we "try a bunch" (which is just binary proto" then "just assume it's proto"

Arguably we could flat out fail to load things with .yaml extensions but given that the legacy path will try to load .csv as json and output "can not load foo.csv as json" I a path saying "can't load foo.yaml as proto" is comparable (and IMO keeps the code simpler not having even more failure paths but you're welcome to disagree and I'll refactor)

Reply:
Because of the principle of least surprise, I think it would be a good idea to make the handling of this method a bit more consistent. I'd suggest:

if `.pb` treat the file as binary proto:
  throw if parsing fails.
  return

if `.pb_txt` treat the file as text proto:
  throw if not supported or if parsing fails.
  return

if `.yml` or `.yaml` treat the file as yaml:
  throw if not supported or if parsing fails.
  return

treat the file as json:
  throw if not supported or if parsing fails.
  return

Reply:
I don't disagree but that's a change to Envoy handling in general which I would strongly prefer to not make as part of this PR as anyone with bad file naming will have an unpleasant surprise.

As an alternate, what if I just had a fully guarded noyaml implementation which literally tried binary proto (else fail) and text proto (else fail) and leave the base impl as is for someoe else to clean up?

Reply:
Thanks for the offline discussion and for reworking this method. It's *SO* much easier to understand the semantics now! (of course we still need to address the "silent" fail issues you discovered, but ... Follow up! :>)